bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/ToolBar.java

```diff
@@ -0,0 +1,592 @@
+			redraw();
+		}
+	}
+
+	private void onMouseUp(Event event) {
+		boolean paintRequested = false;
+		List<ToolItem> copy = List.copyOf(items);
+		for (ToolItem item : copy) {
+			paintRequested |= item.notifyMouseUp(toPoint(event));
+		}
+
+		if (paintRequested) {
+			redraw();
+		}
+	}
+
+	private Point toPoint(Event e) {
+		Point p = e.getLocation();
+		if (isLeftToRight()) {
+			return p;
+		} else {
+			return isLeftToRight() ? p : new Point(getBounds().width - p.x, p.y);
+		}
+	}
+
+	private void onPaint(Event event) {
+		if (!isVisible()) {
+			return;
+		}
+
+		Rectangle bounds = getBounds();
+		if (bounds.width == 0 && bounds.height == 0) {
+			return;
+		}
+
+		Drawing.drawWithGC(this, event.gc, gc -> renderer.render(gc, bounds));
+	}
+
+	/**
+	 * Returns the preferred size (in points) of the receiver.
+	 * <p>
+	 * The <em>preferred size</em> of a control is the size that it would best be
+	 * displayed at. The width hint and height hint arguments allow the caller to
+	 * ask a control questions such as "Given a particular width, how high does the
+	 * control need to be to show all of the contents?" To indicate that the caller
+	 * does not wish to constrain a particular dimension, the constant
+	 * <code>SWT.DEFAULT</code> is passed for the hint.
+	 * </p>
+	 * <p>
+	 * If the changed flag is <code>true</code>, it indicates that the receiver's
+	 * <em>contents</em> have changed, therefore any caches that a layout manager
+	 * containing the control may have been keeping need to be flushed. When the
+	 * control is resized, the changed flag will be <code>false</code>, so layout
+	 * manager caches can be retained.
+	 * </p>
+	 *
+	 * @param wHint   the width hint (can be <code>SWT.DEFAULT</code>)
+	 * @param hHint   the height hint (can be <code>SWT.DEFAULT</code>)
+	 * @param changed <code>true</code> if the control's contents have changed, and
+	 *                <code>false</code> otherwise
+	 * @return the preferred size of the control.
+	 *
+	 * @exception SWTException
+	 *                         <ul>
+	 *                         <li>ERROR_WIDGET_DISPOSED - if the receiver has been
+	 *                         disposed</li>
+	 *                         <li>ERROR_THREAD_INVALID_ACCESS - if not called from
+	 *                         the thread that created the receiver</li>
+	 *                         </ul>
+	 *
+	 * @see Layout
+	 * @see #getBorderWidth
+	 * @see #getBounds
+	 * @see #getSize
+	 * @see #pack(boolean)
+	 * @see "computeTrim, getClientArea for controls that implement them"
+	 */
+	@Override
+	public Point computeSize(int wHint, int hHint, boolean changed) {
+		checkWidget();
+		int zoom = getZoom();
+		wHint = (wHint != SWT.DEFAULT ? DPIUtil.scaleUp(wHint, zoom) : wHint);
+		hHint = (hHint != SWT.DEFAULT ? DPIUtil.scaleUp(hHint, zoom) : hHint);
+		return DPIUtil.scaleDown(computeSizeInPixels(wHint, hHint, changed), zoom);
+	}
+
+	Point computeSizeInPixels(int wHint, int hHint, boolean changed) {
```

Comment:
If I am not mistaken, these lines have to be removed. The calculation done by the renderer is performed in points, thus treating the result as pixels leads to wrong results. Simply treating it as points (without any scaling in between) seem to work fine for me.
```suggestion
		return DPIUtil.scaleDown(computeSizeInPixels(wHint, hHint, changed), zoom);
	}

	Point computeSizeInPixels(int wHint, int hHint, boolean changed) {
```